lib/elixir/lib/file.ex

```diff
@@ -1027,43 +1027,50 @@ defmodule File do
         IO.read(file, :line)
       end)
 
+  See `open/2` for a list of `modes`.
   """
   @spec open(Path.t, [mode | :ram], (io_device -> res)) :: {:ok, res} | {:error, posix} when res: var
-  def open(path, modes, function) do
+  def open(path, modes, function) when is_list(modes) and is_function(function, 1) do
     case open(path, modes) do
-      {:ok, device} ->
+      {:ok, io_device} ->
         try do
-          {:ok, function.(device)}
+          {:ok, function.(io_device)}
         after
-          :ok = close(device)
+          :ok = close(io_device)
         end
       other -> other
     end
   end
 
   @doc """
-  Same as `open/2` but raises an error if file could not be opened.
+  Similar to `open/2` but raises an error if file could not be opened.
 
-  Returns the `io_device` otherwise.
+  Returns the IO device otherwise.
+
+  See `open/2` for a list of `modes`.
   """
-  @spec open!(Path.t, [mode]) :: io_device | no_return
-  def open!(path, modes \\ []) do
+  @spec open!(Path.t, [mode | :ram]) :: io_device | no_return
+  def open!(path, modes \\ []) when is_list(modes) or is_function(modes) do
```

Comment:
I'm not getting that comment. All Elixir tests pass if I use `is_function(modes, 1)` here and in the other cases. 😕